Demos/Cube/Cube.cpp

```diff
@@ -0,0 +1,220 @@
+}
+
+void Cube::timer_event(Core::TimerEvent&)
+{
+    const Vec3f vertices[8] {
+        { -1, -1, -1 },
+        { -1, 1, -1 },
+        { 1, 1, -1 },
+        { 1, -1, -1 },
+        { -1, -1, 1 },
+        { -1, 1, 1 },
+        { 1, 1, 1 },
+        { 1, -1, 1 },
+    };
+
+#define QUAD(a, b, c, d) a, b, c, c, d, a
+
+    const int indices[] {
+        QUAD(0, 1, 2, 3),
+        QUAD(7, 6, 5, 4),
+        QUAD(4, 5, 1, 0),
+        QUAD(3, 2, 6, 7),
+        QUAD(1, 5, 6, 2),
+        QUAD(0, 3, 7, 4),
+    };
+
+    const Color colors[] {
+        Color::Red,
+        Color::Red,
+        Color::Green,
+        Color::Green,
+        Color::Blue,
+        Color::Blue,
+        Color::Magenta,
+        Color::Magenta,
+        Color::White,
+        Color::White,
+        Color::Yellow,
+        Color::Yellow,
+    };
+
+    Vec3f transformed_vertices[8];
+
+    static float angle = 0;
+    angle += 0.02f;
+
+    auto matrix = Mat4x4f::translate(Vec3f(0, 0, 1.5f))
+        * Mat4x4f::rotate(Vec3f(1, 0, 0), angle * 1.17356641f)
+        * Mat4x4f::rotate(Vec3f(0, 1, 0), angle * 0.90533273f)
+        * Mat4x4f::rotate(Vec3f(0, 0, 1), angle);
+
+    for (int i = 0; i < 8; i++) {
+        transformed_vertices[i] = matrix.transform_point(vertices[i]);
+    }
+
+    GUI::Painter painter(*bitmap);
+    painter.fill_rect_with_gradient(Gfx::Orientation::Vertical, painter.clip_rect(), Gfx::Color::White, Gfx::Color::Blue);
+
+    auto to_point = [](const Vec3f& v) {
+        return Gfx::Point(v.x, v.y);
+    };
+
+    for (size_t i = 0; i < sizeof(indices) / sizeof(indices[0]) / 3; i++) {
+        auto a = transformed_vertices[indices[i * 3]];
+        auto b = transformed_vertices[indices[i * 3 + 1]];
+        auto c = transformed_vertices[indices[i * 3 + 2]];
+        auto normal = (b - a).cross(c - a);
+        normal.normalize();
+
+        // Perspective projection
+        a.x = WIDTH / 2 + a.x / (1 + a.z * 0.35) * WIDTH / 3;
+        a.y = HEIGHT / 2 - a.y / (1 + a.z * 0.35) * WIDTH / 3;
+        b.x = WIDTH / 2 + b.x / (1 + b.z * 0.35) * WIDTH / 3;
+        b.y = HEIGHT / 2 - b.y / (1 + b.z * 0.35) * WIDTH / 3;
+        c.x = WIDTH / 2 + c.x / (1 + c.z * 0.35) * WIDTH / 3;
+        c.y = HEIGHT / 2 - c.y / (1 + c.z * 0.35) * WIDTH / 3;
+
+        float winding = (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
+        if (winding < 0)
+            continue;
+
+        float shade = 0.5f + normal.y * 0.5f;
+        auto color = colors[i];
+        color.set_red(color.red() * shade);
+        color.set_green(color.green() * shade);
+        color.set_blue(color.blue() * shade);
+
+        painter.draw_triangle(to_point(a), to_point(b), to_point(c), color);
+    }
+
+    if ((cycles % 50) == 0) {
+        dbgprintf("%d total cycles. finished 50 in %d ms, avg %d ms\n", cycles, timeAvg, timeAvg / 50);
+        stats->set_text(String::format("%d ms", timeAvg / 50));
+        timeAvg = 0;
+    }
+
+    update();
+}
+
+/*
+ * Main
+*/
```

Comment:
Not sure what this adds beyond +3 line count but ok.

Reply:
Hey I stole this from the Fire Demo :D